### PR TITLE
PID1 Horizon Mode Fixed

### PR DESCRIPTION
**Pull request by borisbstyle**
Due to this request:
cleanflight#840

I ported the logic from PID2 into PID1 for the horizon code with modifications for integer instead of floats. Also Level D term is here being used for Horizon Sensitivity.
Default value of 100 equals to 80% sensitivity, which should be fine!

Also the horizon leveling is shifted by 4 to make the default Level I term more usable.

It has been flight tested and it works really well!

### CODE REVIEW

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -659,6 +659,36 @@ static void pidRewrite(pidProfile_t *pidProfile, controlRateConfig_t *controlRat
...
+        horizonLevelStrength = (500 - mostDeflectedPos) / 5;  // 100 at centre stick, 0 = max stick deflection
+
+        // PID D Level Term is used for Horizon Sensitivity. It is adjusted so the default value of 100 works pretty well.
+        // Default Level D term of 100 equals to 80% sensitivity and 125 and above is 100% sensitivity
+        if(pidProfile->D8[PIDLEVEL] == 0){
+            horizonLevelStrength = 0;
+        } else if (pidProfile->D8[PIDLEVEL] >= 125){
+        	horizonLevelStrength = 100;
+        } else {
+            horizonLevelStrength = constrainf(((horizonLevelStrength - 100) * (125 / pidProfile->D8[PIDLEVEL])) + 100, 0, 100);
```
Comment (ledvinap) [inline src/main/flight/pid.c]: 125/D8[] will be truncated ... 
```
horizonLevelStrength  = constrainf((125 * (horizonLevelStrength - 100) / pidProfile->D8[PIDLEVEL])) + 100, 0, 100);
```

Comment (ledvinap) [inline src/main/flight/pid.c]: and `constrain` is enough, no need for float version

**Comment by borisbstyle** [comment]
Thanks for the feedback. Yeah the f slipped away as I copied it from luxfloat logic. Will correct that.

You are right about D8. I actually was using fixed value for that before and wanted to make it configurable. The 125 needs to be 1250 there to get 80.

**Comment by borisbstyle** [comment]
Ok....final fix to make configuring through level D gain possible now.

**Comment by ledvinap** [comment]
Did someone test it?

**Comment by borisbstyle** [comment]
@ledvinap it probably may be simplified in that way, but this was what I flight tested and it works really well. 
Just try it and it will work exactly like pid2 or pid5 horizon. It is tested by my and someone else who I let try.

Pid1 had no horizon at all.....well just a very bad angle mode. D term wasn't being used at all and it is not a real D.
Level P = angle strength
Level I = horizon strength
Level D(new) = horizon sensitivity....(transition from horizon to level)

This will be documented in manuals of course.

**Comment by ledvinap** [comment]
As currently coded:
D8[PIDLEVEL] == 125 -> maximum leveling, not influenced by stick position
D8[PIDLEVEL] == 124 -> leveling based on stick position, maximum in center, none at extreme
D8[PIDLEVEL] == 100 -> leveling based on stick position, maximum in center, 20% at extreme
D8[PIDLEVEL] == 1 -> maximum leveling, not influenced by stick position
D8[PIDLEVEL] == 0 -> no leveling

**Comment by borisbstyle** [comment]
It definately works, but I did not test/project the extreme values towards 1 and 124. Should have projected the curve to notice that behaviour with value 1 and 124 for example.

Thanks for another sharp notice. I will think again about this implementation and perhaps think about removing level D at all or implement it inverted like people would expect from D as it usually works that way.

I don't mind that its in the master for now as this works much better than the regular horizon or better said no horizon mode at all in the current pid1.

But damn I missed floats in here :)

**Comment by ledvinap** [comment]
@borisbstyle : What is intended behavior ?

I got the impression that this value should mark stick deflection beyond which leveling influence is zero ...

**Comment by borisbstyle** [comment]
I see now.

It is all as expected, except the highest value 125 should not be like that.
125 still needs to go through stick position logic and it's all fine.

I made a little think error when I wanted to limit the max value of 125.

will adjust that

**Comment by borisbstyle** [comment]
Ok this should be fine, see code below:
D125 min stick=100 max stick= 0, center stick=50
D124  min stick=100 max stick= 1, center stick=50
D100 min stick=100 max stick= 20, center stick=60
D60 min stick=100, max stick=52, center stick=76
D10 min stick=100, max stick=92, center stick=96
D1 min stick=100, max stick=100, center stick=100
D0 min stick=100, max stick=100, center stick=100


if (pidProfile->D8[PIDLEVEL] >= 125){
        	horizonLevelStrength = constrain((horizonLevelStrength - 100) + 100, 0, 100);
        } else {
        	horizonLevelStrength = constrain((10 * (horizonLevelStrength - 100) * (10 * pidProfile->D8[PIDLEVEL] / 125) / 100) + 100, 0, 100);
}

**Comment by borisbstyle** [comment]
https://github.com/cleanflight/cleanflight/pull/895

**Comment by ledvinap** [comment]
It may make sense to allow higher D value (use lower divisor instead of 125). That way there will be stick position beyond which no leveling is applied (pure ACRO mode)

**Comment by borisbstyle** [comment]
Ok good idea.....I will test it with higher values and test the behaviour.

**Comment by borisbstyle** [comment]
The reason I use 125 is to make the default value of 100 as the most optimal setting as 80-90% is what works the best according to some people.

**Comment by MasterZap** [comment]
Man, if this keeps up, we may have Boris flyign Horizon mode :)

#justkidding

**Comment by borisbstyle** [comment]
Haha...I now can understand why people coming from level would use this mode.
I think the latest patch made it perfect. I updated the cc3d hex from the link above.

**Pull request merged**